### PR TITLE
client authorization

### PR DESCRIPTION
**Pull request by billfort**
Signed-off-by: billfort <fxbao@hotmail.com>

### CODE REVIEW

File: website/sidebars.json [hunk 1]
```diff
@@ -21,7 +21,8 @@
         "nkn-sdk-js",
         "nkn-sdk-go",
         "nkn-sdk-gomobile",
-        "nkn-sdk-java"
+        "nkn-sdk-java",
+        "client-authorization"
```
Comment (yilunzhang) [inline website/sidebars.json]: Should be in `protocol` section

File: docs/client-authorization.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+When a NKN client joins NKN network, it sends a RPC request to seed list to enquiry which node that it should connect by webscoket.
+
+After get the webscoket node infomation, the client connects to the node, and send its client Id, public key to the node.
+
+Then the node need check that the client is the right client who owns this client Id and public. This is client authorization for.
+
+The main steps are here:
+1. client connect to webscoket node
+2. node generate 256 bit random data as a <font color=red>challenge</font> , and send to client
```
Comment (yilunzhang) [inline docs/client-authorization.md]: This is markdown not html...

File: docs/client-authorization.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
+---
+id: client-authorization
+title: NKN Client Authorization
+---
+
+When a NKN client joins NKN network, it sends a RPC request to seed list to enquiry which node that it should connect by webscoket.
+
+After get the webscoket node infomation, the client connects to the node, and send its client Id, public key to the node.
```
Comment (yilunzhang) [inline docs/client-authorization.md]: get -> getting
send -> sends

File: docs/client-authorization.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+---
+id: client-authorization
+title: NKN Client Authorization
+---
+
+When a NKN client joins NKN network, it sends a RPC request to seed list to enquiry which node that it should connect by webscoket.
+
+After get the webscoket node infomation, the client connects to the node, and send its client Id, public key to the node.
+
+Then the node need check that the client is the right client who owns this client Id and public. This is client authorization for.
```
Comment (yilunzhang) [inline docs/client-authorization.md]: need -> needs to
This is client authorization for -> This is what client authorization is for

File: website/sidebars.json (hunk 1 above)
Comment (billfort) [reply website/sidebars.json]: fixed

File: docs/client-authorization.md (hunk 2 above)
Comment (billfort) [reply docs/client-authorization.md]: yes. removed this html tag.

File: docs/client-authorization.md (hunk 3 above)
Comment (billfort) [reply docs/client-authorization.md]: fixed

File: docs/client-authorization.md (hunk 4 above)
Comment (billfort) [reply docs/client-authorization.md]: fixed

File: docs/client-authorization.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+When a NKN client joins NKN network, it sends a RPC request to seed list to inquire which node that it should connect by webscoket.
+
+After getting the websocket node information, the client connects to the node, and sends its client Id, public key to the node.
+
+Then the node needs to check that the client is authorized to use this **client address** . This is what client authorization is for.
+
+The main steps are here:
+1. A client connects to a node by webscoket;
+2. When getting a new connection from a client, the node generates 256 bit data as a **challenge** , and send it to the client. This action is *authChallenge*;
```
Comment (yilunzhang) [inline docs/client-authorization.md]: might be good to say it's random data in 2 and 3

File: docs/client-authorization.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+3. After getting the challenge data, the client generates 256 bit data as **client salt**, appends it to **challenge** data, and get **hash** 256 of it;
```
Comment (yilunzhang) [inline docs/client-authorization.md]: **hash** 256 -> sha256

File: docs/client-authorization.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+4. The client signs the **hash** by its private key, gets a **signature**;
+5. The client sends the **client salt**, **signature** and client address to the node. This action is *setClient*;
+6. The node gets public key from the client address;
+6. The node verifies the **signature** by the client public key;
+7. If the verification is successful, the node keeps the connection, else the node **disconnect** the client.
```
Comment (yilunzhang) [inline docs/client-authorization.md]: disconnects

File: docs/client-authorization.md (hunk 5 above)
Comment (billfort) [reply docs/client-authorization.md]: Yes. fixed.

File: docs/client-authorization.md (hunk 6 above)
Comment (billfort) [reply docs/client-authorization.md]: yes. fixed

File: docs/client-authorization.md (hunk 7 above)
Comment (billfort) [reply docs/client-authorization.md]: yes. fixed.

**Comment by yilunzhang** [comment]
Looks good to me. Let's merge this after client authorization code is merged

**Pull request merged**